### PR TITLE
[WIP] Netgen Integration

### PR DESCRIPTION
**Pull request by UZerbinati**

### CODE REVIEW

**Review by dham — CHANGES_REQUESTED** [review]
This is lovely and very nearly good to go. Can we please just have one of the tests set to run in parallel just to ensure that that works (and stays working).

**Comment by UZerbinati** [comment]
Sorry,it turns out that making it work in parallel is more complicated then expected :( This comes from the fact that NetGen already stores the Mesh in parallel. I exposed the `DMPlexBuildFromCellListParallel` method in petsc4py ([https://gitlab.com/UZerbinati1/petsc](url) the fork) to try solving this issue, but I'm having some trouble using it (on going discussion in the Slack).

File: tests/regression/test_netgen.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                while sum_marked_eta < part*sum_eta:
+                    new_marked = (~marked) & (eta > frac*eta_max)
+                    sum_marked_eta += sum(eta[new_marked])
+                    marked += new_marked
+                    frac -= delfrac
+                markedVec0.getArray()[:] = 1.0*marked[:]
+            sct(markedVec0,markedVec,mode=PETSc.Scatter.Mode.REVERSE)
+            return markedVec
+
+    def Refine(msh, marked):
```
Comment (dham) [inline tests/regression/test_netgen.py:175]: make this a mesh method. Have `marked` be a DG0 field over the mesh which is 0 or 1.

File: tests/regression/test_netgen.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        w = TestFunction(W)
+        R_T = lam.real*uh + div(grad(uh))
+        n = FacetNormal(V.mesh())
+        h = CellDiameter(msh)
+        R_dT = dot(grad(uh), n)
+        eta = assemble(h**2*R_T**2*w*dx + (h("+")+h("-"))*(R_dT("+")-R_dT("-"))**2*(w("+")+w("-"))*dS)
+        frac = .95
+        delfrac = .05
+        part = .2
+        with eta.dat.vec as etaVec:
```
Comment (dham) [inline tests/regression/test_netgen.py:152]: It would be much nicer to do this in Firedrake.

File: firedrake/mesh.py [hunk 3]
```diff
@@ -249,6 +254,75 @@ def facet_cell_map(self):
...
+    TODO: Right now we construct NetGen mesh on a single worker, load it in Firedrake
+    and then distribute. We should find a way of taking adventage of the fact that
+    NetGen can act as a parallel mesher.
+    """
+    if ngmesh.dim == 3:
+        if comm.Get_rank() == 0:
+            V = ngmesh.Coordinates()
+            T = ngmesh.Elements3D().NumPy()["nodes"]
+            T = np.array([list(np.trim_zeros(a, 'b')) for a in list(T)])-1
+            surfMesh, dim = False, 3
```
Comment (UZerbinati) [inline firedrake/mesh.py:272]: write dim as tdim

File: firedrake/mesh.py [hunk 4]
```diff
@@ -249,6 +254,75 @@ def facet_cell_map(self):
...
+                surfMesh, dim = True, 2
+                T = ngmesh.Elements2D().NumPy()["nodes"]
+                T = np.array([list(np.trim_zeros(a, 'b')) for a in list(T)])-1
+            plex = PETSc.DMPlex().createFromCellList(dim, T, V, comm=comm)
+            plex.setName(_generate_default_mesh_topology_name(ngmesh.GetMeshName()))
+            vStart, vEnd = plex.getDepthStratum(0)
+            if surfMesh:
+                for e in ngmesh.Elements1D():
+                    join = plex.getJoin([vStart+v.nr-1 for v in e.vertices])
+                    plex.setLabelValue(dmcommon.FACE_SETS_LABEL, join[0], int(e.surfaces[1]))
```
Comment (UZerbinati) [inline firedrake/mesh.py:283]: EDGE_SETS_LABEL not FACE_SETS_LABEL

File: firedrake/mesh.py [hunk 5]
```diff
@@ -249,6 +254,75 @@ def facet_cell_map(self):
...
+            else:
+                for e in ngmesh.Elements2D():
```
Comment (UZerbinati) [inline firedrake/mesh.py:285]: Add python comment on why this should not give performance problem

File: firedrake/mesh.py [hunk 6]
```diff
@@ -2106,7 +2182,35 @@ def Mesh(meshfile, **kwargs):
...
+                        if mark[getIdx(i)]:
+                            el.SetRefinementFlag(1)
+                        else:
+                            el.SetRefinementFlag(0)
+                    self.netgen_mesh.RefineFlaged(0, True)
+                    return Mesh(self.netgen_mesh)
+                else:
+                    return Mesh(netgen.libngpy._meshing.Mesh(2))
+
+        setattr(MeshGeometry, "Refine", Refine)
```
Comment (UZerbinati) [inline firedrake/mesh.py:2212]: Check if `setattr` is needed.

File: firedrake/mesh.py [hunk 7]
```diff
@@ -2106,7 +2182,35 @@ def Mesh(meshfile, **kwargs):
...
+                    getIdx = lambda x: x
+                    _, marked0 = self.topology_dm.distributeField(self.sfBCInv, self._cell_numbering, marked)
+                if self.comm.Get_rank() == 0:
+                    mark = marked0.getArray()
+                    for i, el in enumerate(self.netgen_mesh.Elements2D()):
+                        if mark[getIdx(i)]:
+                            el.SetRefinementFlag(1)
+                        else:
+                            el.SetRefinementFlag(0)
+                    self.netgen_mesh.RefineFlaged(0, True)
```
Comment (UZerbinati) [inline firedrake/mesh.py:2207]: Correct spelling

File: scripts/firedrake-install [hunk 8]
```diff
@@ -1791,7 +1807,7 @@ if mode == "install":
         # When building Firedrake in CI, don't clone Firedrake again.
         check_call(["ln", "-s", "../../firedrake", "firedrake"])
     else:
-        git_clone("git+https://github.com/firedrakeproject/firedrake.git")
+        git_clone("git+https://github.com/firedrakeproject/firedrake.git@uzerbinati/netgen-integration")
```
Comment (UZerbinati) [inline scripts/firedrake-install:1810]: No need

File: scripts/firedrake-install [hunk 9]
```diff
@@ -1410,6 +1414,16 @@ def build_and_install_slepc():
             git_clone(url)
 
 
+def build_and_install_netgen():
+    run_pip_install(["scikit-build"])
+    log.info("Installing Netgen.")
+    log.info("Depending on your platform, NetGen may take an hour or more to build!")
```
Comment (UZerbinati) [inline scripts/firedrake-install:1420]: Add optional dep on netgen setup.py. Add flag to install netgen with OCC.

File: tests/regression/test_netgen.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+from netgen.occ import *
+import netgen
+import numpy as np
+import gc
+from petsc4py import PETSc
+
+
+printf = lambda msg: PETSc.Sys.Print(msg)
+
+
```
Comment (UZerbinati) [inline tests/regression/test_netgen.py:13]: Easier test. Integrating function on netgen mesh of square, cube and sphere.

File: tests/regression/test_netgen.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
+from firedrake import *
+from mpi4py import MPI
+from netgen.geom2d import SplineGeometry
+from netgen.occ import *
```
Comment (UZerbinati) [inline tests/regression/test_netgen.py:4]: not import all, choose what to import

File: tests/regression/test_netgen.py [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+import gc
+from petsc4py import PETSc
+import pytest
+
+
+printf = lambda msg: PETSc.Sys.Print(msg)
+
+
+def poisson(h, degree=2):
+    comm = MPI.COMM_WORLD
```
Comment (JDBetteridge) [inline tests/regression/test_netgen.py:18]: COMM_WORLD is already in the firedrake namespace (you get it from from firedrake import *)

File: tests/regression/test_netgen.py [hunk 13]
```diff
@@ -0,0 +1,269 @@
+from firedrake import *
+from mpi4py import MPI
```
Comment (JDBetteridge) [inline tests/regression/test_netgen.py:2]: See comment below, this import can be removed

File: scripts/firedrake-install [hunk 14]
```diff
@@ -427,6 +429,8 @@ else:
                         help="Cleans up build artefacts from Firedrake venv. Unless you really cannot spare the disk space, this option is not recommended")
     parser.add_argument("--with-blas", default=None,
                         help="Specify path to system BLAS directory. Use '--with-blas=download' to download openblas")
+    parser.add_argument("--netgen", action="store_true", dest="netgen", default=config["options"]["netgen"],
+                        help="Install Netgen as well.")
```
Comment (dham) [inline scripts/firedrake-install:433]:
```suggestion
                        help="Install Netgen.")
```

**Review by dham — DISMISSED** [review]

File: scripts/firedrake-install [hunk 15]
```diff
@@ -323,6 +323,8 @@ honoured.""",
                         help="Cleans up build artefacts from Firedrake venv. Unless you really cannot spare the disk space, this option is not recommended")
     parser.add_argument("--with-blas", default=None,
                         help="Specify path to system BLAS directory. Use '--with-blas=download' to download openblas")
+    parser.add_argument("--netgen", action="store_true",
+                        help="Install Netgen as well.")
```
Comment (dham) [inline scripts/firedrake-install:327]:
```suggestion
                        help="Install Netgen.")
```

**Review by dham — APPROVED** [review]

**Pull request merged**